entities/schema/crossref/bulk_builder.go

```diff
@@ -0,0 +1,118 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2023 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package crossref
+
+import (
+	"fmt"
+
+	"github.com/go-openapi/strfmt"
+)
+
+// BulkBuilder is a low-alloc tool to build many beacon strings (as []byte). It
+// is optimized to allocate just once as opposed to once per ID. This makes it
+// considerably faster when generating 100s of thousand of beacons strings. The
+// main intended use case for this is building propValuePairs in ref-filters.
+//
+// The BulkBuilder makes some estimations for how much memory will be necessary
+// based on expected input params. If those requirements get exceeded, it will
+// still be safe to use, but will fallback to allocating dynamically.
+type BulkBuilder struct {
+	buf    []byte
+	prefix []byte
+	offset int
+}
+
+func NewBulkBuilderWithEstimates(expectedCount int, exampleClassName string,
+	overheadRatio float64,
+) *BulkBuilder {
+	prefix := []byte("weaviate://localhost/")
+
+	lenOfTypicalClassName := int(float64(len(exampleClassName)) * overheadRatio)
+	predictedSize := expectedCount * (len(prefix) + 1 + lenOfTypicalClassName + 36)
+
+	bb := &BulkBuilder{
+		buf:    make([]byte, predictedSize),
+		prefix: prefix,
+	}
+
+	return bb
+}
+
+func (bb *BulkBuilder) ClassAndID(className string,
+	id strfmt.UUID,
+) []byte {
+	requiredSpace := len(bb.prefix) + len(className) + 1 + len(id)
+	if bb.offset+requiredSpace > len(bb.buf) {
+		return bb.fallbackWithClassName(className, id)
+	}
+
+	// copy the start pos, we will need this at the end to know what to return to
+	// the caller
+	start := bb.offset
```

Comment:
this looks like the perfect place for the byte operations package ;)

you'd need to add a WriteUint8 function, but otherwise it should have everything

Reply:
done, much cleaner indeed. No need for a `WriteUint8` because we already had `WriteByte` and Go considers them the same.